### DISCUSSION TITLE
Generating private keys

### DISCUSSION THREAD

**Discussion by leopashov** [MetaMask/snaps#1832] [General]
I am trying to get to grips with handling private keys within metamask snaps. My aim is to derive the account keys I get when using metamask flask (ie create a snap which will return my account private and public key as a string, similar to the 'hello' example. I have been using polygon to do this using the BIP44 and BIP32 methods as described at https://docs.metamask.io/snaps/reference/rpc-api/. with coinType: 966 and path: ['m', "44'", "966'"], curve: 'secp256k1' for BIP44 and BIP32 methods respectively, but they are different keys (I feel I may be lacking some general understanding here but expected them to be the same). I've also started looking at the 'snap_getEntropy' call, but can't figure out how to derive a private key from the entropy. How is this done with metamask? In each of these methods of creating new private keys, is the mnemonic being used? Is there a way of building a snap to access the mnemonic?
Code at https://github.com/leopashov/snap

**Comment by Mrtenz** [comment]
I had a look at your code, and it looks like your BIP-32 and BIP-44 methods are using different derivation paths. For BIP-44, you're using:

```ts
const deriveAddress = await getBIP44AddressKeyDeriver(addressNode);
const secondAccount = deriveAddress(0);
```

which will derive the key at `address_index` `0` from the parent node.

For BIP-32, you're using:

```ts
const addressNode = await snap.request({
  // [...]
  path: ['m', "44'", "966'"],
});

// [...]
const accountKey0 = await slip10Node.derive(["bip32:0'"]);
```

which will derive the child `0'` of the root node. These are two different derivation paths. BIP-44 derivation paths look like this:

```
m / purpose' / coin_type' / account' / change / address_index
```

So in the case of BIP-44, you're deriving with a `coin_type` of `966'` and `address_index` of `0`, meaning `m / 44' / 966' / 0' / 0 / 0`.  For BIP-32, you're using `m / 44' / 966'` as the root node, and deriving `0'` from it, resulting in `m / 44' / 966' / 0'`. To get the same keys using BIP-32, you can just change the initial path to `['m', "44'", "966'", "0'", '0', '0']` (or the parent, and then derive the child node at index `0`).

I hope that helps!

**Reply by leopashov** [reply]
That sorted it, Thank you very much!
Just wanted to ask regarding derivation paths - To what extent are they specific to a given chain? 
It seems metamask generates addresses using the m/44'/60'/... path but these addresses are consistent across EVM chains?
Doesn't this negate the need for separate chains having separate derivation paths (or coin types)?

**Reply by Mrtenz** [reply]
Typically you want to have a different address for each chain. Other chains like Bitcoin use `m / 44' / 1'` instead for example. There is a list of `coin_type`s here: https://github.com/satoshilabs/slips/blob/master/slip-0044.md

**Reply by Montoya** [reply]
There's a whole rabbit hole here for you if you are interested! m/44'/60' is what Ethereum used and within the Ethereum ecosystem there started to be a trend of other chains that are EVM compliant also using the same derivation path, so accounts that worked on Ethereum could also work on Polygon, BSC, etc. Also, various layer 2 networks like Linea, Optimism, Arbitrum, etc. choose to use the same address format and so wallets just use the same derivation path for these networks as well. It's a design decision within Ethereum.

**Reply by leopashov** [reply]
Thank you very much! This makes a lot more sense now!
